indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt

```diff
@@ -0,0 +1,284 @@
+            val result: CredentialDefinition? = null
+
+            repeat(RETRY_TIMES) {
+                try {
+                    val getCredDefRequest = Ledger.buildGetCredDefRequest(did, credDefId).get()
+                    val getCredDefResponse = Ledger.submitRequest(pool, getCredDefRequest).get()
+                    val credDefIdInfo = Ledger.parseGetCredDefResponse(getCredDefResponse).get()
+
+                    return@runBlocking SerializationUtils.jSONToAny<CredentialDefinition>(credDefIdInfo.objectJson)
+                } catch (e: Exception) {
+                    logger.debug("Credential definition retrieving failed (id: $credDefId). Retry attempt $it")
+                    delay(DELAY_MS)
+                }
+            }
+
+            result
+        }
+
+        /**
+         * Retrieves revocation registry definition from ledger
+         *
+         * @param did           indy user did
+         * @param pool          indy pool handle
+         * @param revRegDefId   target revocation registry definition id
+         *
+         * @return              revocation registry definition or null if none exists on ledger
+         */
+        fun retrieveRevocationRegistryDefinition(did: String?, pool: Pool, revRegDefId: String): RevocationRegistryDefinition? = runBlocking {
+            val result: RevocationRegistryDefinition? = null
+
+            repeat(RETRY_TIMES) {
+                try {
+                    val request = Ledger.buildGetRevocRegDefRequest(did, revRegDefId).get()
+                    val response = Ledger.submitRequest(pool, request).get()
+                    val revRegDefJson = Ledger.parseGetRevocRegDefResponse(response).get().objectJson
+
+                    return@runBlocking SerializationUtils.jSONToAny<RevocationRegistryDefinition>(revRegDefJson)
+                } catch (e: Exception) {
+                    logger.debug("Revocation registry definition retrieving failed (id: $revRegDefId). Retry attempt $it")
+                    delay(DELAY_MS)
+                }
+            }
+
+            result
+        }
+
+        /**
+         * Retrieves revocation registry entry from ledger
+         *
+         * @param did           indy user did
+         * @param pool          indy pool handle
+         * @param revRegId      revocation registry id
+         * @param timestamp     time from unix epoch in seconds representing time moment you are interested in
+         *                      e.g. if you want to know current revocation state, you pass 'now' as a timestamp
+         *
+         * @return              revocation registry entry or null if none exists on ledger
+         */
+        fun retrieveRevocationRegistryEntry(did: String?, pool: Pool, revRegId: String, timestamp: Int): Pair<Int, RevocationRegistryEntry>? = runBlocking {
+            val result: Pair<Int, RevocationRegistryEntry>? = null
```

Comment:
I slightly concerned about both approaches. Let's investigate best practice in Corda. It may be restricted to sleep during the flow execution.

Reply:
This subject should be investigated more carefully. This function is called from some corda flows, so it shouldn't operate with threads to be correctly serialized with Kryo. Coroutine will work with 100% probability because it's just a function with a context.

Reply:
> Coroutine will work with 100% probability

No, `runBlocking` blocks a thread the same way `Thread.sleep` does. AFAIU Quasar, it is okey to sleep on a thread since any Java function may do this.